### PR TITLE
Intermediate cache clearing

### PR DESCRIPTION
**Pull request by jayqi**
Addresses #12 

- Adds method `hasCompletedAllDownstreamModules` to `WorkflowInterface` and `DAGWorkflow` classes to check whether a given module's downstream modules have all completed.
- Adds intermediate cache removal to `Execute`. Upon completion of a module, it will evaluate `hasCompletedAllDownstreamModules` on all its upstream modules and remove the upstream module's cache if true.

### CODE REVIEW

**Comment by jayqi** [comment]
Note that this implementation may sometimes fail to clear a module's cache during execution. For example, suppose you had this workflow:
```
A -> B -> C
 ↘
  D
```
If module C completes, then it will clear module B's cache. But if this happens before module D completes, then `hasCompletedAllDownstreamModules(A)` will always evaluate to FALSE because B has already been cleared. 

This is because `module$hasCompleted()` checks for the existence of the cache to determine completion. We'd need to maintain global states of completion at the workflow level in order to have complete, correct information.

**Comment by jayqi** [comment]
@cwschultz88 in case you missed it. ^

**Comment by jameslamb** [comment]
@jayqi I really appreciate the way you ASCII-arted that example. Super clear

File: tests/testthat/test_workflow_dag.R [hunk 1]
```diff
@@ -122,7 +126,7 @@ test_that("Testing DAGWorkflow Class Static initFromFile Method", {
     })
 })
 
-test_that("Testing DAGWorkflow obj methods error when appriopriate", {
+test_that("Testing DAGWorkflow obj methods error when appropriate", {
```
Comment (cwschultz88) [inline tests/testthat/test_workflow_dag.R:129]: nice catch

File: R/workflow_dag.R [hunk 2]
```diff
@@ -337,42 +344,52 @@ DAGWorkflow <- R6::R6Class("DAGWorkflow"
...
-        
+
         , getDownstreamModules = function(module) {
             return(private$getNeighbors(private$graph
                                         , module
                                         , "out"))
         }
-        
+
+        , hasCompletedAllDownstreamModules = function(module) {
```
Comment (cwschultz88) [inline R/workflow_dag.R:366]: Glad to see this method only took a few lines

**Review by cwschultz88 — APPROVED** [review]
Everything looks great, Thanks Jay!

**Pull request merged**